File: 2020 카카오 인턴십/수식 최대화.java
import java.util.*;

class Solution {
    long answer = Long.MIN_VALUE;
    
    List<Long> numList = new ArrayList<>();
    List<String> operList = new ArrayList<>();
    
    String[] oper = { "+", "-", "*" };
    String[] output = new String[3];
    boolean[] visited = new boolean[3];
    
    long cal(long n1, long n2, String o) {
        long res = 0;
        switch(o) {
            case "+":
                res = n1 + n2;
             	break;
            case "-":
                res = n1 - n2;
             	break;
            case "*":
                res = n1 * n2;
             	break;
        }
        return res;
    }
    
    void solve() {
        List<String> oper = new ArrayList<>();
        oper.addAll(operList);
        
        List<Long> num = new ArrayList<>();
        num.addAll(numList);
        
        for(int i = 0; i < output.length; ++i) {
            String curOper = output[i];
            for(int j = 0; j < oper.size(); ++j) {
                if(oper.get(j).equals(curOper)) {
                    long n1 = num.get(j);
                    long n2 = num.get(j+1);
                    long res = cal(n1, n2, curOper);
                    
                    num.remove(j+1);
                    num.remove(j);
                    oper.remove(j);
                    
                    num.add(j, res);
                    j--;
                }
            }
        }
        answer = Math.max(answer, Math.abs(num.get(0)));
    }
    
    void per(int depth, int r) {
        if(depth == r) {
            solve();
            return;
        }
        
        for(int i = 0; i < oper.length; ++i) {
            if(!visited[i]) {
                visited[i] = true;
                output[depth] = oper[i];
                per(depth+1, r);
                visited[i] = false;
            }
        }
    }
    
    public long solution(String expression) {
        String n = "";
        for(int i = 0; i < expression.length(); ++i) {
            char exp = expression.charAt(i);
            if(exp == '+' || exp == '-' || exp == '*') {
                operList.add(exp + "");
                numList.add(Long.parseLong(n));
                n = "";
            }
            else n += exp;
        }
        numList.add(Long.parseLong(n));
        per(0, oper.length);
        return answer;
    }
}

File: 2020 카카오 인턴십/수식 최대화.cpp
#include <iostream>
#include <string>
#include <vector>
#include <cmath>
#include <algorithm>

using namespace std;

long long answer;
vector<long long> num, _num;
vector<char> oper, _oper;

string allOper = "*+-";
bool check[3];

void getNumAndOper(string exp) {
    string tmp = "";
    for(int i = 0; i < exp.length(); ++i) {
        if(exp[i] == '*' || exp[i] == '+' || exp[i] == '-') {
            num.push_back(stoll(tmp));
            tmp = "";
            oper.push_back(exp[i]);
            continue;
        }
        tmp += exp[i];
        if(i == exp.length()-1)
            num.push_back(stoll(tmp));
    }
}

long long calc(char op, long long n1, long long n2) {
    if(op == '*') return n1*n2;
    if(op == '+') return n1+n2;
    if(op == '-') return n1-n2;
}

long long getResult(string myOper) {
    _num = num, _oper = oper;
    for(char op : myOper)
        for(int i = 0; i < _oper.size(); ++i) 
            if(_oper[i] == op) {
                long long tmp = calc(_oper[i], _num[i], _num[i+1]);
                _oper.erase(_oper.begin()+i);
                _num.erase(_num.begin()+i);
                _num.erase(_num.begin()+i);
                _num.insert(_num.begin()+i, tmp);
                i -= 1;
            }
    return _num.front();
}

void setPriority(int cnt, string myOper) {
    if(cnt == 3) {
        answer = max(answer, abs(getResult(myOper)));
        return;
    }
    for(int i = 0; i < 3; ++i) {
        if(check[i] == false) {
            check[i] = true;
            setPriority(cnt+1, myOper+allOper[i]);
            check[i] = false;
        }
    }
}

long long solution(string expression) {
    getNumAndOper(expression);
    setPriority(0, "");
    return answer;
}
